packages/govuk-frontend/src/govuk/components/accordion/accordion.test.js

```diff
@@ -733,6 +733,37 @@ describe('/components/accordion', () => {
               message: 'GOV.UK Frontend is not supported in this browser'
             })
           })
+
+          it('throws when $module is not set', async () => {
+            await expect(
+              renderAndInitialise(page, 'accordion', {
+                params: examples.default,
+                beforeInitialisation($module) {
+                  $module.remove()
+                }
+              })
+            ).rejects.toEqual({
+              name: 'MissingElementError',
+              message: 'Accordion: $module not found'
+            })
+          })
+
+          it('throws when receiving the wrong type for $module', async () => {
+            await expect(
+              renderAndInitialise(page, 'accordion', {
+                params: examples.default,
+                beforeInitialisation($module) {
+                  // Replace with an `<svg>` element which is not an `HTMLElement` in the DOM (but an `SVGElement`)
+                  $module.outerHTML = `<svg data-module="${$module.getAttribute(
+                    'data-module'
+                  )}"></svg>`
```

Comment:
Yo know how we clearly put `accordion` on line 753?

Can we help newcomers out and save them a click by doing the same here?

```suggestion
                  $module.outerHTML = '<svg data-module="govuk-accordion"></svg>'
```

Reply:
Keeps the [simplicity of how you did it initially](https://github.com/alphagov/govuk-frontend/pull/4104#pullrequestreview-1617358822) above? I loved that

```js
$svg.setAttribute('data-module', 'accordion')
```

Together with `$module` we can be both simple and obvious

Reply:
I appreciate that it's more all in plain sight to have an hardcoded `data-module="govuk-accordion"`. However given this test is the same from one component to another, I think there's a benefit for grabbing it from the existing element [^1]. That quickly allows porting it to a new component without risking to put the wrong module name in the attribute, one place less to worry about.

As mentionned before, I also think it clarifies what happens: we're not replacing with an element with an arbitrary `data-module`, we're injecting an element with the same `data-module` as the one that was there. 



I think there are other solutions if we're worried about that line being clearer to newcommers:
1. Add extra comment to clarify that we're pulling the component's attribute to ensure the initialisation will happen on the replacement element:

```suggestion
                  // Replace with an `<svg>` element which is not an `HTMLElement` in the DOM (but an `SVGElement`)
				  // reusing the `data-module` attribute of the original element to ensure the component 
                  // will be instantiated on this replacement element
                  $module.outerHTML = `<svg data-module="${$module.getAttribute(
                    'data-module'
                  )}"></svg>`
```

2. That line is a bit unwieldly being on two line, but it could easily be made more readable by getting the value earlier:

```suggestion
                  // Grab `data-module` from the existing element for reuse in the replacement element
                  // to ensure the component will initialise on it
                  const moduleName = $module.getAttribute(
                    'data-module'
                  );
                  
                  // Replace with an `<svg>` element which is not an `HTMLElement` in the DOM (but an `SVGElement`)
                  $module.outerHTML = `<svg data-module="${moduleName}"></svg>`
```

@alphagov/design-system-developers From discussing on Slack, we don't seem to be able to agree on a way to go there, would anyone be able to weigh in, please?

[^1]: I also acknowledge that the matter would be less of a question should that value have been the selector we use to grab the element under test, or a value compared in an expectation, where putting that hardcoded string straight in the tests would be much safer in the long run.

Reply:
I don't really have a strong opinion either way, but given we seem to be at a bit of an impasse, I'd lean towards hardcoding the attribute.

There are good reasons on both sides, but my thinking is:

- there are two other places in the block of code that would be copied where the component name is already hardcoded, so when porting to a new component you already have to change them anyway
- there are fewer things to go wrong (e.g. `data-module` being `undefined` for some unanticipated reason – with a hardcoded attribute there are fewer uncertainties you have to reason about in the code)
- we should prefer self-documenting code where possible, and the fact that we think this might need a comment is a bit of a 'code smell' for me

Having said all of that, we should bear in mind that this is entirely internal and easy to change – I don't think I would personally make this a blocking thing if I was reviewing.

I do think it'd be neat to abstract the tests for common component functionality into a reusable set of tests at some point – is it worth creating an issue to do that once we've got v5 out the way?

Reply:
Cheers for weighing on this. I've hardcoded the attributes in my last push so that should be all clear for a last review. I'll raise an issue for abstracting tests for common components functionnality as well.